go/glome/glome.go

```diff
@@ -0,0 +1,228 @@
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package glome implements GLOME protocol.
+package glome
+
+import (
+	"fmt"
+	"io"
+
+	"crypto/hmac"
+	"crypto/sha256"
+
+	"golang.org/x/crypto/curve25519"
+)
+
+const (
+	// PrivateKeySize is the size of a PrivateKey in bytes.
+	PrivateKeySize = 32
+	// PublicKeySize is the size of a PublicKey in bytes.
+	PublicKeySize = 32
+	// MaxTagSize is the maximum size allowed for a Tag
+	MaxTagSize = 32
+	// MinTagSize is the minimum size allowed for a Tag
+	MinTagSize = 1
+)
+
+var (
+	// ErrInvalidPublicKey denotes that a slice that intend to be a public key is not of desired length
+	ErrInvalidPublicKey = fmt.Errorf("Invalid public key - byte slice len is not %d", PublicKeySize)
+	// ErrInvalidPrivateKey denotes that a slice that intend to be a private key is not of desired length
+	ErrInvalidPrivateKey = fmt.Errorf("Invalid private key - byte slice len is not %d", PrivateKeySize)
+	// ErrInvalidTagSize denotes that provided integer is not suitable to be minPeerTagSize
+	ErrInvalidTagSize = fmt.Errorf("Invalid tag size - minPeerTagSize must be in range %d-%d",
+		MinTagSize, MaxTagSize)
+	// ErrInvalidReader denotes that library failed to read PrivateKeySize bytes from given Reader.
+	ErrInvalidReader = fmt.Errorf("Invalid Reader - Failed to read  %d bytes", PrivateKeySize)
+)
+
+// PublicKey is the type of GLOME public Keys.
+//
+// It can be initializes either by casting a [PublicKeySize]byte array or from a byte
+// slice with the PublicKeyFromSlice function. For example:
+//
+//     //Generate Private Key as existing byte array
+//     b := [32]byte{0,2,...,7,6}
+//     p := glome.PublicKey(b)
+//     //Generate from byte slice
+//     s := b[:]
+//     p, _ := glome.PublicKeyFromSlice(s)
+type PublicKey [PublicKeySize]byte
+
+// PublicKeyFromSlice generates a PublicKey object from slice. Return ErrInvalidPublicKey
+// if slice's length is not PublicKeySize.
+func PublicKeyFromSlice(b []byte) (*PublicKey, error) {
+	if len(b) != PublicKeySize {
+		return nil, ErrInvalidPublicKey
+	}
+
+	var p PublicKey
+	copy(p[:], b)
+	return &p, nil
+}
+
+// PrivateKey is the type of GLOME public Keys.
+//
+// It can be initializes either by casting a [PrivateKeySize]byte array or from a byte
+// slice with the PrivateKeyFromSlice function. For example:
+//
+//     //Generate Private Key as existing byte array
+//     b := [32]byte{0,2,...,7,6}
+//     p := glome.PrivateKey(b)
+//     //Generate from byte slice
+//     s := b[:]
+//     p, _ := glome.PrivateKeyFromSlice(s)
```

Comment:
So I remain unconvinced that this is a great API design. For the exchange I'd optimally want to do something like this:

```go
eph, err := glome.GenerateKey()
if err != nil { [...] }
// for glome-login we actually need an identifier, too
pub, err := ioutil.ReadFile("/etc/glome/service.pub")
if err != nil { [...] }
srv, err := glome.PublicKey(pub)
if err != nil { [...] }
tag, err := eph.Exchange(srv).Tag(msg)
if err != nil { [...] }
```

And the other side:

```go
// for glome-login we actually need an identifier, too
p, err := ioutil.ReadFile("/etc/glome/service.key")
if err != nil { [...] }
priv, err := glome.PrivateKey(p)
if err != nil { [...] }
// get peer public key
tag, err := priv.Exchange(peer).Tag(msg)
if err != nil { [...] }
```

It's definitely not ok to drop the error on the floor in the example when the whole point of the function is to do the length check. If you want to use byte arrays they should be an implementation detail.

Reply:
The API that we have now is really similiar to the one that you are requesting. For the exchange the actual code can be:
```go
_, eph, err := glome.GenerateKey(rand.Reader)
if err != nil { [...] }
// for glome-login we actually need an identifier, too
pub, err := ioutil.ReadFile("/etc/glome/service.pub")
if err != nil { [...] }
srv, err := glome.PublicKeyFromSlice(pub)
if err != nil { [...] }
tag, err := eph.Exchange(srv).Tag(msg)
if err != nil { [...] }
```

And the other side:
```go
// for glome-login we actually need an identifier, too
p, err := ioutil.ReadFile("/etc/glome/service.key")
if err != nil { [...] }
priv, err := glome.PrivateKeyFromSlice(p)
if err != nil { [...] }
// get peer public key
tag, err := priv.Exchange(peer).Tag(msg)
if err != nil { [...] }
```

All the difference are:
1. GenerateKey need a source of entropy. 
2. GenerateKey also returns the public Key - this can be change if it is considered that there will be no use for public key in most cases
3. `s/glome.PrivateKey(p)/glome.PrivateKeyFromSlice(p)` and
`s/glome.PublicKey(p)/glome.PublicKeyFromSlice(p)`

To address everyone of them:
1. This was decided in order not to impose the user which source of entropy to use. As there is no optional parameter in golang, to use a default allowing `nil` as a valid input was considered. It was discarded because it was regarded as unnecesary if how to generate it in a standard way (with `crypto/rand`) is conveniently documented.
2. I though that having a PublicKey from the begginning might be useful, but it is really a question of taste. Totally fine with changing this option if it is considered necessary
3. There is no difference between the sample and the actual code but the name of a function. We can discuss alternative for the naming.


Finally, I recall that this is not a great example by any means. Will change it to be more useful, both in documentation and in the README.

Reply:
Discussed offline, let's move on by calling this current version of the API alpha and subject to change.

@pedrobn23 can you add a note somewhere in the README of the library that this is Alpha and the API is not fixed yet?